pkg/binlog/binlog.go

```diff
@@ -0,0 +1,294 @@
+	return c.nodeStatus(ctx, "drainers")
+}
+
+func (c *Client) nodeID(ctx context.Context, addr, ty string) (string, error) {
+	nodes, err := c.nodeStatus(ctx, ty)
+	if err != nil {
+		return "", err
+	}
+
+	addrs := []string{}
+	for _, node := range nodes {
+		if addr == node.Addr {
+			return node.NodeID, nil
+		}
+		addrs = append(addrs, addr)
+	}
+
+	return "", errors.Errorf("%s node id for address %s not found, found address: %s", ty, addr, addrs)
+}
+
+// UpdateDrainerState update the specify state as the specified state.
+func (c *Client) UpdateDrainerState(ctx context.Context, addr string, state string) error {
+	nodeID, err := c.nodeID(ctx, addr, "drainers")
+	if err != nil {
+		return err
+	}
+	return c.updateStatus(ctx, "drainers", nodeID, state)
+}
+
+// UpdatePumpState update the specify state as the specified state.
+func (c *Client) UpdatePumpState(ctx context.Context, addr string, state string) error {
+	nodeID, err := c.nodeID(ctx, addr, "pumps")
+	if err != nil {
+		return err
+	}
+	return c.updateStatus(ctx, "pumps", nodeID, state)
+}
+
+// updateStatus update the specify state as the specified state.
+func (c *Client) updateStatus(ctx context.Context, ty string, nodeID string, state string) error {
+	key := fmt.Sprintf("/tidb-binlog/v1/%s/%s", ty, nodeID)
+
+	resp, err := c.etcdClient.KV.Get(ctx, key)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errors.Errorf("no pump with node id: %v", nodeID)
+	}
+
+	var nodeStatus NodeStatus
+	err = json.Unmarshal(resp.Kvs[0].Value, &nodeStatus)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	if nodeStatus.State == state {
+		return nil
+	}
+
+	nodeStatus.State = state
+
+	data, err := json.Marshal(&nodeStatus)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	_, err = c.etcdClient.Put(ctx, key, string(data))
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	return nil
+}
+
+func (c *Client) nodeStatus(ctx context.Context, ty string) (status []*NodeStatus, err error) {
+	key := fmt.Sprintf("/tidb-binlog/v1/%s", ty)
+
+	resp, err := c.etcdClient.KV.Get(ctx, key, clientv3.WithPrefix())
+	if err != nil {
+		return nil, errors.AddStack(err)
+	}
+
+	for _, kv := range resp.Kvs {
+		var s NodeStatus
+		err = json.Unmarshal(kv.Value, &s)
+		if err != nil {
+			return nil, errors.Annotatef(err, "key: %s,data: %s", string(kv.Key), string(kv.Value))
+		}
+
+		status = append(status, &s)
+	}
+
+	return
+}
+
+func (c *Client) offline(addr string, nodeID string) error {
+	url := c.getOfflineURL(c.hookAddr(addr), nodeID)
+	req, err := http.NewRequest("PUT", url, nil)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
+		return errors.Errorf("error requesting %s, code: %d",
+			resp.Request.URL, resp.StatusCode)
+	}
+
+	defer resp.Body.Close()
```

Comment:
Do we need to move this `Close` before `if resp.StatusCode < 200 || resp.StatusCode >= 400 {`?